### PR TITLE
Generate null safe build scripts if possible

### PR DESCRIPTION
**Pull request by simolus3**
Migrates:

- the builders of `package:build_test`
- `build_vm_compilers`
- the internal integration tests in `_test`

Then, the build script generation is changed to emit null safe scripts if all builder imports and the root package are opted in to null safety.

### CODE REVIEW

File: build_runner/lib/src/build_script_generate/build_script_generate.dart [hunk 1]
```diff
@@ -109,11 +133,71 @@ Future<Iterable<Expression>> findBuilderApplications({
...
+      featureSet: FeatureSet.fromEnableFlags2(
+          sdkLanguageVersion: version, flags: const []),
+      throwIfDiagnostics: false,
+    );
+    if (parsedFile.errors.isNotEmpty) {
+      // Don't try to assume a language version if the imported file is invalid
+      return false;
+    }
+
+    final definedLanguageVersion = parsedFile.unit.languageVersionToken;
```
Comment (jakemac53) [inline build_runner/lib/src/build_script_generate/build_script_generate.dart:184]: Instead of manually doing this detection based on versions, we should be able to do something like:

```
return parsedFile.unit.featureSet.isEnabled(Feature.non_nullable);
```

**Comment by jakemac53** [comment]
> I think we need a more complex test for fully null-safe builds, but I'm not sure where to put them? Into `build_script_generate_test.dart`?

I think we should just move the majority of integration tests to run in null safety mode. That might just happen automatically as a result of this PR. We don't need to heavily test the opted out version.

**Comment by jakemac53** [comment]
> So far, there are other issues in `build_runner` preventing sound null safety (encountered by building the example):

Did you want to look at those as a part of this pr? Or want one of us to look at them and push fixes here?

File: build_runner/lib/src/build_script_generate/build_script_generate.dart [hunk 2]
```diff
@@ -181,13 +180,7 @@ Future<bool> _allMigratedToNullSafety(PackageGraph packageGraph,
...
     }
 
-    final definedLanguageVersion = parsedFile.unit.languageVersionToken;
-    if (definedLanguageVersion != null) {
-      version = Version(
-          definedLanguageVersion.major, definedLanguageVersion.minor, 0);
-    }
-
-    if (version < _minVersionForNullSafety) return false;
+    return parsedFile.unit.featureSet.isEnabled(Feature.non_nullable);
```
Comment (jakemac53) [inline build_runner/lib/src/build_script_generate/build_script_generate.dart:183]: I realized I gave you the wrong code here, :)
```suggestion
    if (!parsedFile.unit.featureSet.isEnabled(Feature.non_nullable)) return false;
```

**Comment by simolus3** [comment]
> Did you want to look at those as a part of this pr?

It turned out to be a very minor issue, so I've just fixed it now :D I also migrated the `build_vm_compilers` package, that might enable null safe integration tests.

The markdown check fails because the HN reader has been removed from the angulardart repo, do you know a good example that could replace it?

**Comment by simolus3** [comment]
I think the failing tests are only due to the deleted example from angular and because there's no analyzer version for 2.14.

**Comment by jakemac53** [comment]
> The markdown check fails because the HN reader has been removed from the angulardart repo, do you know a good example that could replace it?

I don't, I just removed it in another of my PRs.

**Comment by jakemac53** [comment]
> I also migrated the `build_vm_compilers` package, that might enable null safe integration tests.

Fwiw, that package is essentially deprecated (although not officially yet). It has very few usages, and even fewer motivating use cases left.

The test package will soon be using the frontend_server to compile tests which will make it basically always faster for running tests than build_vm_compilers (in reality it usually was anyways due to the unrelated overhead in build_runner). This was the primary original use case, and it is no longer relevant.

The only remaining use case really then is to run apps which have generated code in them, without having to do that from a generated `build` dir. We can support that separately (we can keep the `build_runner run` command), by creating a merged dir under the covers, or using http uris in a custom package config pointing at a dev server.

**Comment by jakemac53** [comment]
@simolus3 can  you try merging with master and re-running the checks?

**Comment by jakemac53** [comment]
Another test fix pending in https://github.com/dart-lang/build/pull/3082

**Comment by simolus3** [comment]
> Fwiw, that package is essentially deprecated (although not officially yet). It has very few usages, and even fewer motivating use cases left.

That's good to know, thanks! I always liked using that package on large CLI apps to improve startup times during development (just keep a `build_runner watch` running and start from kernel). Is that a valid use case that has another solution now?

Either way, I mainly migrated that package to make the integration tests run with null safety :)

**Comment by jakemac53** [comment]
> Is that a valid use case that has another solution now?

It is a valid use case - but not one that we have many users actually utilizing that we are aware of. There is also a better option available (but not implemented by any tools afaik) - using the frontend server to incrementally recompile. This would make the dev experience essentially that of flutter. This could also support hot reload and hot restart.

It would be cool to build something into the `dart` tool which could do this...

File: build_vm_compilers/CHANGELOG.md [hunk 3]
```diff
@@ -1,6 +1,7 @@
 ## 1.0.9-dev
 
 - Update to build version `2.0.0`.
+- Migrate to null safety
```
[the remark this answers is not in the document]
Comment (simolus3) [reply build_vm_compilers/CHANGELOG.md]: It can compile files with null safety, but the generated dills don't run with sound null safety (unless I'm missing a flag). I've added a small note to clarify this.

File: _test_common/lib/descriptors.dart [hunk 4]
```diff
@@ -19,14 +19,19 @@ import 'package:test_descriptor/test_descriptor.dart' as d;
...
-    Map<String, String> pathDependencies = const {},
-    Map<String, String> versionDependencies = const {}}) async {
+///
+/// The [sdkEnvironment] describes the `environment.sdk` field in the pubspec.
+Future<d.FileDescriptor> pubspec(
+  String name, {
+  Iterable<String> currentIsolateDependencies = const [],
+  Map<String, String> pathDependencies = const {},
+  Map<String, String> versionDependencies = const {},
+  String sdkEnvironment = '>=2.9.0 <3.0.0',
```
[the remark this answers is not in the document]
Comment (simolus3) [reply _test_common/lib/descriptors.dart:29]: I'll follow the CI run at https://github.com/simolus3/build/pull/4 to see how many things it breaks :)

[the remark this answers is not in the document]
Comment (simolus3) [reply _test_common/lib/descriptors.dart]: Looks like it works without further changes

File: build_vm_compilers/CHANGELOG.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jakemac53) [reply build_vm_compilers/CHANGELOG.md]: Ah ya its the sound/unsound distinction that it doesn't actually support then. I expected initially to need to do some additional stuff around the package config but actually build_modules is likely doing this.

**Comment by jakemac53** [comment]
Technically there is one failure scenario here, somebody could have a legacy import hidden transitively somewhere, and we wouldn't detect it, and then it would fail to run in sound mode even though the builder imports were opted in.

In an ideal world I would say we would detect a failure to run in sound mode, log a warning or something, and then fall back to unsound mode. But, I don't think that is really worth doing in practice?

I think we can take a wait and see approach here, if people file issues relating to this we can add a flag to opt out of the behavior, or a fallback to unsound mode.

**Review by jakemac53 — APPROVED** [review]

**Comment by jakemac53** [comment]
thanks again @simolus3 !
